api/database/functions.py

```diff
@@ -1,31 +1,103 @@
+                                 Token, stgReport)
 from fastapi import HTTPException
 from sqlalchemy import text
 from sqlalchemy.exc import InternalError, OperationalError
-from sqlalchemy.sql.expression import select, insert
-from datetime import datetime, timedelta
+from sqlalchemy.sql.expression import insert, select
+from sqlalchemy.sql.sqltypes import TIMESTAMP
 
 logger = logging.getLogger(__name__)
 
+
 def list_to_string(l):
     string_list = ', '.join(str(item) for item in l)
     return string_list
-    
+
+
+async def is_valid_rsn(rsn: str) -> bool:
+    return re.fullmatch('[\w\d\s_-]{1,13}', rsn)
+
+
+async def to_jagex_name(name: str) -> str:
+    return name.lower().replace('_', ' ').replace('-', ' ').strip()
+
+
+async def jagexify_names_list(names: List[str]) -> List[str]:
+    return [await to_jagex_name(n) for n in names if await is_valid_rsn(n)]
+
+
+async def sql_select_players(names):
+    sql = select(Player)
+    sql = sql.where(Player.normalized_name.in_(tuple(await jagexify_names_list(names))))
+    async with get_session(EngineType.PLAYERDATA) as session:
+        data = await session.execute(sql)
+    data = sqlalchemy_result(data)
+    return [] if not data else data.rows2dict()
+
+
+async def sql_insert_player(new_names):
+    sql = insert(Player)
+    async with get_session(EngineType.PLAYERDATA) as session:
+        await session.execute(sql, new_names)
+        await session.commit()
+
+
+async def sql_insert_report(param):
+    sql = insert(stgReport)
+    async with get_session(EngineType.PLAYERDATA) as session:
+        await session.execute(sql, param)
+        await session.commit()
+
+
+async def parse_detection(data: dict) -> dict:
+    gmt = time.gmtime(data['ts'])
+    human_time = time.strftime('%Y-%m-%d %H:%M:%S', gmt)
+
+    equipment = data.get('equipment', {})
+
+    param = {
+        'reportedID': data.get('id'),
+        'reportingID': data.get('reporter_id'),
+        'region_id': data.get('region_id'),
+        'x_coord': data.get('x_coord'),
+        'y_coord': data.get('y_coord'),
+        'z_coord': data.get('z_coord'),
+        'timestamp': human_time,
+        'manual_detect': data.get('manual_detect'),
+        'on_members_world': data.get('on_members_world'),
+        'on_pvp_world': data.get('on_pvp_world'),
+        'world_number': data.get('world_number'),
+        'equip_head_id': equipment.get('equip_head_id'),
+        'equip_amulet_id': equipment.get('equip_amulet_id'),
+        'equip_torso_id': equipment.get('equip_torso_id'),
+        'equip_legs_id': equipment.get('equip_legs_id'),
+        'equip_boots_id': equipment.get('equip_boots_id'),
+        'equip_cape_id': equipment.get('equip_cape_id'),
+        'equip_hands_id': equipment.get('equip_hands_id'),
+        'equip_weapon_id': equipment.get('equip_weapon_id'),
+        'equip_shield_id': equipment.get('equip_shield_id'),
+        'equip_ge_value': data.get('equip_ge_value', 0)
+    }
+    return param
```

Comment:
this does not belong in the functions.py & should be cleaned up

Reply:
It's a series of functions, if not in functions then it belongs in either a report_functions.py file, as reports.py seems to be for routes and not logic/cleanup relates.